# Learning Numpy

## Theory

### Import Numpy

In [2]:
import numpy as np

### Create basic array

In [8]:
zero_array_30 = np.zeros(30)
zero_array_30

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [9]:
np.ones(30)

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [10]:
np.empty(30)

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [15]:
np.arange(2,10,3)

array([2, 5, 8])

In [16]:
py_list = [1,2,3,4,5]
np.array(py_list)

array([1, 2, 3, 4, 5])

### Basic Functions

In [17]:
monthly_closings_aapl_2020 = np.array([100,102,103,101,96,111,113,116,120,150,140,134])
monthly_closings_aapl_2021 = np.array([136,144,138,150,180,142,145,150,165,177,165,155])

In [18]:
monthly_closings_aapl_2021.shape

(12,)

In [19]:
np.sort(monthly_closings_aapl_2020)

array([ 96, 100, 101, 102, 103, 111, 113, 116, 120, 134, 140, 150])

With .reshape() we can change the shape of an array.
Note: New Array needs to have same number of elements as original array.


In [21]:
quarterly_closing_aapl_2020 = monthly_closings_aapl_2020.reshape(4,3) #quarterly 
quarterly_closing_aapl_2020

array([[100, 102, 103],
       [101,  96, 111],
       [113, 116, 120],
       [150, 140, 134]])

In [23]:
quarterly_closing_aapl_2020[0, 2]

np.int64(103)

### Slicing

In [25]:
# Get Closing Prices < 105
monthly_closings_aapl_2020[monthly_closings_aapl_2020 > 120]

array([150, 140, 134])

In [28]:
# Get Closing Prices > 120 & < 130
monthly_closings_aapl_2020[(monthly_closings_aapl_2020 > 100) & (monthly_closings_aapl_2020 < 130)]

array([102, 103, 101, 111, 113, 116, 120])

### Math Functions

In [29]:
monthly_closings_aapl_2020.sum()

np.int64(1386)

In [30]:
monthly_closings_aapl_2020.min()

np.int64(96)

In [31]:
monthly_closings_aapl_2020.max()

np.int64(150)

In [32]:
# broadcasting
monthly_closings_aapl_2020 * 0.8

array([ 80. ,  81.6,  82.4,  80.8,  76.8,  88.8,  90.4,  92.8,  96. ,
       120. , 112. , 107.2])

In [38]:
np.random.uniform(-3,3, 10)

array([ 1.48403421, -0.23312404, -1.77320279, -0.71467116,  2.86720715,
        0.77621129,  2.34463485,  0.05312005,  2.71872601,  1.28961382])

## Excercise 1
### NumPy Exercise: Portfolio Analysis and Risk Management

### Learning Objectives
- Master NumPy array creation, indexing, and slicing
- Apply NumPy operations to financial data
- Implement functions with input validation
- Combine NumPy with control flow structures

---

### Exercise 1: Stock Portfolio Analysis
#### Background
You are a portfolio analyst managing a portfolio of 5 stocks over 10 trading days. You need to analyze the daily prices and calculate various performance metrics.

#### Task 1.1: Create the Portfolio Data

Create a 2D NumPy array called `portfolio_prices` with shape (10, 5) representing:
- **Rows**: 10 trading days
- **Columns**: 5 different stocks (Stock A, B, C, D, E)

Use the following initial prices for Day 0:
- Stock A: 100, Stock B: 50, Stock C: 75, Stock D: 120, Stock E: 30

Generate random daily returns between -3% and +3% for each subsequent day using `np.random.uniform()`.

**Hint**: Use cumulative products to simulate realistic price movements.
```python

In [56]:
all_returns

array([[1.        , 1.        , 1.        , 1.        , 1.        ],
       [0.99247241, 1.02704286, 1.01391964, 1.00591951, 0.97936112],
       [0.97935967, 0.97348502, 1.02197057, 1.0060669 , 1.01248435],
       [0.97123507, 1.02819459, 1.01994656, 0.98274035, 0.9809095 ],
       [0.98100427, 0.98825453, 1.00148539, 0.9959167 , 0.98747375],
       [1.00671117, 0.97836963, 0.98752868, 0.99198171, 0.9973642 ],
       [1.01711056, 0.98198043, 1.00085407, 1.00554487, 0.97278702],
       [1.00645269, 0.98023145, 0.9739031 , 1.02693313, 1.02793792],
       [1.01850384, 0.98827683, 0.97586033, 1.01105398, 0.99640915],
       [0.97732229, 0.99971061, 0.97206331, 1.02455922, 0.9855268 ]])

In [4]:
np.random.seed(42)

initial_prices = np.array([100, 50, 75, 120, 30])

daily_returns = np.random.uniform(0.97, 1.03, size=(9, 5))

all_returns = np.vstack([np.ones((1, 5)), daily_returns])

portfolio_prices = initial_prices * all_returns

print(portfolio_prices.shape)
print(portfolio_prices)

(10, 5)
[[100.          50.          75.         120.          30.        ]
 [ 99.24724071  51.35214292  76.04397274 120.71034109  29.38083355]
 [ 97.93596712  48.67425084  76.64779266 120.72802808  30.37453064]
 [ 97.12350697  51.40972956  76.49599188 117.9288416   29.42728494]
 [ 98.10042706  49.41272673  75.11140394 119.51000413  29.62421245]
 [100.67111737  48.91848158  74.06465092 119.03780527  29.92092597]
 [101.71105577  49.09902135  75.06405497 120.6653849   29.18361074]
 [100.64526911  49.01157237  73.04273217 123.23197587  30.83813766]
 [101.85038409  49.41384131  73.18952451 121.32647779  29.89227449]
 [ 97.73222941  49.98553073  72.90474835 122.94710689  29.56580397]]


---

#### Task 1.2: Basic Indexing and Slicing

Using the `portfolio_prices` array, extract the following:

a) The prices of Stock C on all days  
b) The prices of all stocks on Day 5  
c) The prices of Stocks B, C, and D on Days 3 to 7 (inclusive)  
d) The last closing price of each stock
```python

In [18]:
stock_c_prices = portfolio_prices[:,2]
stock_prices_day_5 = portfolio_prices[5,:]
stock_prices_b_c_d_day_3_7 = portfolio_prices[3:8,1:4]
stock_prices_last = portfolio_prices[-1,:]
stock_prices_last

array([ 97.73222941,  49.98553073,  72.90474835, 122.94710689,
        29.56580397])

---

#### Task 1.3: Calculate Daily Returns

Write a function `calculate_returns(prices)` that:
- Takes a 2D array of prices (shape: days × stocks)
- **Validates** that the input is a 2D NumPy array with at least 2 rows
- Calculates daily percentage returns for each stock
- Returns a 2D array with shape (days-1, stocks)

**Formula**: Return = (Price_today - Price_yesterday) / Price_yesterday * 100
```python